KAFKA-14226 (KIP-821): [connect:transform] Introduce support for nested structures

jeqo opened the pull request:
[KIP-821](https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures) implementation.

Includes the following changes:

- Introduces a new modules inside `o.a.k.connect.transforms`: `fields`
- FieldPath interface and 2 implementations: `SingleFieldPath` and `MultiFieldPaths`
- Utils to hold and update Map, Struct and Schema values.
- Add support for nested fields to the following SMTs:
  - ExtractField
  - HeaderFrom
  - MaskField
  - TimestampConverter

> Added some SMTs to the same PR to show how the FieldPath is used, instead of separate PRs that will make this harder to appreciate.

Upcoming SMTs will be updated in following PRs to complete the scope of KIP-821.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

yashmayya commented on connect/api/src/main/java/org/apache/kafka/connect/transforms/Transformation.java:
```diff
@@ -24,20 +24,31 @@
+    String FIELD_SYNTAX_VERSION_CONFIG = "field.syntax.version";
+    String FIELD_SYNTAX_VERSION_DOC = "Defines the version of the syntax to access fields. "
+        + "If set to `V1`, then the field paths are limited to access the elements at the root level of the struct or map."
+        + "If set to `V2`, the syntax will support accessing nested elements. o access nested elements, "
```
```suggestion
        + "If set to `V2`, the syntax will support accessing nested elements. To access nested elements, "
```

yashmayya commented on connect/api/src/main/java/org/apache/kafka/connect/transforms/Transformation.java:
```diff
@@ -24,20 +24,31 @@
+        + "dotted notation is used. If dots are already included in the field name, then backtick pairs "
+        + "can be used to wrap field names containing dots. "
+        + "e.g. to access elements from a struct/map named \"foo.bar\", "
+        + "the following format can be used to access its elements: \"`foo.bar`.baz\".";
```
I think this should be `from a field in a struct/map named ...`  instead?

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java:
```diff
@@ -39,33 +41,35 @@
     public static final ConfigDef CONFIG_DEF = new ConfigDef()
-            .define(FIELD_CONFIG, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Importance.MEDIUM, "Field name to extract.");
+            .define(FIELD_CONFIG, ConfigDef.Type.STRING, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Importance.MEDIUM, "Field name to extract.")
+            .define(FIELD_SYNTAX_VERSION_CONFIG, ConfigDef.Type.STRING, FIELD_SYNTAX_VERSION_DEFAULT_VALUE, ConfigDef.Importance.HIGH, FIELD_SYNTAX_VERSION_DOC);
```
Could we add a validator here to ensure that `FIELD_SYNTAX_VERSION_CONFIG` belongs to `FieldSyntaxVersion.values()` (maybe case-insensitive match if desired)?

jeqo replied:
Great idea! adding validator and constants to `FieldSyntaxVersion`

yashmayya commented on connect/api/src/main/java/org/apache/kafka/connect/transforms/Transformation.java:
```diff
@@ -24,20 +24,31 @@
+        + "e.g. to access elements from a struct/map named \"foo.bar\", "
+        + "the following format can be used to access its elements: \"`foo.bar`.baz\".";
+
+    String FIELD_SYNTAX_VERSION_DEFAULT_VALUE = "V1";
```
I think this (and the above config too) can be moved to `FieldPath` or `FieldSyntaxVersion` in the `connect:transforms` subproject since they don't need to be added to the public interface? And then you can avoid using the string form "V1" directly here. 

Or else, if the intention *is* to add it to the public interface and allow custom SMT builders to also use this new notation, then should we consider moving the `FieldPath` and related classes to the `connect:api` subproject?

jeqo replied:
Agree, moving it to `FieldSyntaxVersion`

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+ * <ul>
+ * <li>It follows a dotted notation to represent nested values.</li>
+ * <li>If field names contain dots, can be escaped by wrapping field names with backticks.</li>
+ * <li>If field names contain dots at wrapping positions (beginning or end of path, before or after dots), then backticks need to be
```
```suggestion
 * <li>If field names contain backticks at wrapping positions (beginning or end of path, before or after dots), then backticks need to be
```

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+
+    public static FieldPath ofV2(String field) {
+        return of(field, FieldSyntaxVersion.V2);
+    }
```
nit: can be package-private if only visible for testing

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+                    // if no dots or wrapping backticks are used, then return path with single step
+                    if (!path.contains(DOT)
+                        && !(path.startsWith(BACKTICK) && path.endsWith(
+                        BACKTICK))) {
```
Why do we need this check?

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+                        // prepare for tracking path steps
+                        final List<String> steps = new ArrayList<>();
+                        final StringBuilder s = new StringBuilder(
+                            path); // avoid creating new string on changes
```
```suggestion
                        final StringBuilder s = new StringBuilder(path); // avoid creating new string on changes
```

nit: unnecessary newline

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+                                            s.delete(0, idx + 2); // rm backtick and dot
+                                            break;
+                                        }
+                                    } else { // global backtick
```
Is a "global" backtick referring to a pair of backticks that wrap the whole field path? I didn't get why that requires special handling?

jeqo replied:
Not really. I was trying to optimize for that scenario; though there is not really much to optimize for haha. I removed the concept to avoid confusion.

yashmayya commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,377 @@
+        }
+    }
+
+    private String checkIncompleteBacktickPair(String field) {
```
Would it be possible to add a doc comment for this method? Also `checkIncompleteBacktickPair` seems like a misnomer for a method that is returning a String

jeqo replied:
Good call. I have renamed it to `escapeBackticks` and add docs.

yashmayya commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/FieldPathTest.java:
```diff
@@ -0,0 +1,183 @@
+        assertArrayEquals(new String[] {"foo", "bar\\`.`baz"}, FieldPath.of("foo.`bar\\\\`.\\`baz`", FieldSyntaxVersion.V2).path());
+    }
+
+    @Test void shouldBuildV2WithBackticksWrappingBackticks() {
```
The rules defined in the KIP don't seem to include this case (although there is an example in the subsequent section)? 

https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures#KIP821:ConnectTransformssupportfornestedstructures-Rules

jeqo replied:
Good catch. I will stick to the rules in this case, as there is no need for this special case by using backslash to escape backticks.
Tests are updated.

yashmayya reviewed (commented):
Thanks @jeqo! I've gone through some parts and have some comments and questions.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,232 @@
+     * @param whenNotFound function to apply when current path(s) is/are not found
+     * @return an updated schema. Resulting schemas are usually cached for further access
+     */
+    Schema updateSchemaFrom(
```
This method isn't used. I understand that it may be necessary for a separate PR that implements the rest of the KIP for other SMTs, but to simplify the review process, could we slim this PR down to just the changes that are necessary for the SMTs that are being touched on?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java:
```diff
@@ -0,0 +1,232 @@
+ * @see SingleFieldPath
+ * @see MultiFieldPaths
+ */
+public interface FieldPath {
```
Do we need this interface? It doesn't appear on the left-hand side of any declarations (every SMT that's touched on in this PR explicitly uses either the `SingleFieldPath` or `MultiFieldPaths` class).

jeqo replied:
True. My goal was trying to align the ops between this two implementations.
I think full alignment is achievable, and we can only offer the FieldPath interface to users: https://github.com/jeqo/kafka/pull/4
Let me know what you think. It has some nice side effects described on the internal PR.

C0urante replied:
Hmm... I think the issue with trying to unify around a single `FieldPath` (or `FieldPaths`) interface is that we lose a lot of the guarantees that the Java type system can provide us. For example, in the PR you linked, the implementation of `MultiFieldPaths::fieldFrom` doesn't really guarantee that the user provided a single field, it just operates on a single arbitrarily-selected field, even if the user provided several fields. And I'm not sure of any case where an SMT designed to operate on a single field at a time would prefer to use an internal API designed to operate on several fields instead of a single field (e.g., `FieldsFrom` instead of `FieldFrom`).

Also, even though there's technically nothing preventing users from building against this internal API, it's still not public API and we should not encourage users to build against, e.g., the `FieldPath` interface since it's subject to change at any time, at which point plugins that depend on it would become incompatible with some versions of Kafka Connect.

Ultimately, I don't see us gaining much from this interface right now, since the ways we use the `SingleFieldPath` and `MultiFieldPaths` differ enough (at least as far as the Java type system is concerned) that it's difficult to define a single shared interface for the two of them.

jeqo replied:
Thanks for sharing the reasoning. I agree with your proposal and will be removing the FieldPath interface 👍🏽

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+                    + "If set to `V2`, the syntax will support accessing nested elements. To access nested elements, "
+                    + "dotted notation is used. If dots are already included in the field name, then backtick pairs "
+                    + "can be used to wrap field names containing dots. "
+                    + "e.g. to access elements from a field in a struct/map named \"foo.bar\", "
```
```suggestion
                    + "E.g. to access the subfield `baz` from a field named \"foo.bar\" in a struct/map  "
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+                        FieldSyntaxVersion.FIELD_SYNTAX_VERSION_CONFIG,
+                        ConfigDef.Type.STRING,
+                        FieldSyntaxVersion.FIELD_SYNTAX_VERSION_DEFAULT_VALUE,
+                        FieldSyntaxVersion.FIELD_SYNTAX_VERSION_VALIDATOR,
```
This can be a one-liner, and we can get rid of the custom validator class/field:
```suggestion
                                                ConfigDef.CaseInsensitiveValidString.in(Utils.enumOptions(FieldSyntaxVersion.class)),
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+
+    FieldSyntaxVersion(final String name) {
+        this.name = name;
+    }
```
It looks like we don't need the `name` field for anything. I think we can just use a vanilla, default-constructor enum here.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+                return version;
+            }
+        }
+        throw new ConfigException("Invalid field syntax version");
```
Few things:
1. We can use `Enum::valueOf` with the uppercase name instead of iterating over possible values
2. We can beef up the `ConfigException` to get a slightly nicer error message

```suggestion
        try {
            return FieldSyntaxVersion.valueOf(name.toUpperCase(Locale.ROOT));
        } catch (IllegalArgumentException e) {
            throw new ConfigException(FIELD_SYNTAX_VERSION_CONFIG, name, "Unrecognized field syntax version");
        }
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MapFieldAndValue.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Objects;
+import java.util.StringJoiner;
+
+public final class MapFieldAndValue {
```
Do we need this class? It looks like we could use the [Map.Entry](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html) interface and/or [AbstractMap.SimpleImmutableEntry](https://docs.oracle.com/javase/8/docs/api/java/util/AbstractMap.SimpleImmutableEntry.html) class instead.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/StructFieldAndValue.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Objects;
+import java.util.StringJoiner;
+
+public final class StructFieldAndValue {
```
Similar thought as with the `MapFieldAndValue` class; I think we can replace instances of this class with `Map.Entry<Field, Object>` and/or `AbstractMap.SimpleImmutableEntry<Field, Object>`.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+    public static final String FIELD_SYNTAX_VERSION_DEFAULT_VALUE = V1.name();
+    public static final ConfigDef.Validator FIELD_SYNTAX_VERSION_VALIDATOR = new Validator();
+
+    public static ConfigDef baseConfigDef() {
```
It's slightly more idiomatic to have a `define(ConfigDef baseConfigDef)` or `addToConfigDef(ConfigDef configDef)` (same semantics, just different names) method than to provide a standalone `ConfigDef` for this kind of mix-in logic.

This is because the pattern currently used here doesn't work if we want to add a different kind of mix-in config logic for something else, like handling of not-found fields.

If you want, we could add a method like this to the `Utils` class and then leverage that either here or in calling classes to achieve the same thing:

```java
    public static void addToDefinition(ConfigDef base, ConfigDef toAdd) {
        Objects.requireNonNull(base, "Base ConfigDef may not be null");
        Objects.requireNonNull(toAdd, "Additional ConfigDef may not be null");
        toAdd.configKeys().values().forEach(base::define);
    }
```

jeqo replied:
Good catch. I see also `new ConfigDef(base)` may be cover this already, but that would assume existing transformers are creating ConfigDefs from scratch.
What about adding this method to `ConfigDef` itself?
I gave a try and added `ConfigDef addDefinition(ConfigDef toAdd)` to `ConfigDef`. Let me know what you think

C0urante replied:
`ConfigDef` is public API, we can't make changes to it without a KIP.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        pathTree = buildPathTree(this.paths, 0, new HashMap<>());
+    }
+
+    public static Builder newBuilder(FieldSyntaxVersion syntaxVersion) {
```
Unused method (and the `Builder` class is essentially unused as well)

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+
+    public static MultiFieldPaths of(SingleFieldPath... paths) {
+        return new MultiFieldPaths(Arrays.asList(paths));
+    }
```
These are only used in tests; can we just create a utility method in that test suite to replace these constructors?

I prefer to keep testing-only code inside tests when possible; otherwise, it becomes challenging to read over this stuff if/when debugging is necessary and we only want to look at the code that's actually running in prod

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        return new MultiFieldPaths(Arrays.asList(paths));
+    }
+
+    public static MultiFieldPaths of(List<SingleFieldPath> paths) {
```
Unused method

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        return new MultiFieldPaths(paths);
+    }
+
+    public static MultiFieldPaths of(Set<String> fields, FieldSyntaxVersion syntaxVersion) {
```
Unused method

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    final List<SingleFieldPath> paths;
+
+    MultiFieldPaths(List<SingleFieldPath> paths) {
+        this.paths = paths.stream().filter(Objects::nonNull).collect(Collectors.toList());
```
I'd err on the side of caution here and throw an exception if we see any null elements instead of silently filtering them out.

It's fairly safe to throw from SMTs given the `errors.tolerance` property and the options it gives users on how to handle failures while applying transformations.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+                .collect(Collectors.toList()));
+    }
+
+    Map<String, Object> buildPathTree(List<SingleFieldPath> paths, int stepIdx, Map<String, Object> pathTree) {
```
A few thoughts:

1. Since this is called from the constructor, I'm guessing it can (and probably should) be static? That way, we can test it in isolation in unit tests (if we want to), and it better illustrates the purpose of the method
2. Can we add a Javadoc that states what the purpose of this method is, along with (if applicable) any assumptions about input arguments and invariants about the return value?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+
+/**
+ * Multiple field paths to access data objects ({@code Struct} or {@code Map}) efficiently,
+ * instead of using {@see SingleFieldPath} individually.
```
```suggestion
 * instead of multiple individual {@link SingleFieldPath single-field paths}.
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+ * instead of using {@see SingleFieldPath} individually.
+ * <p>
+ * If the SMT requires accessing a single field on the same data object,
+ * use {@code FieldPath} instead.
```
```suggestion
 * use {@link SingleFieldPath} instead.
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+ * <li>
+ *     <ul>Tree values contain either a nested tree or a field path</ul>
+ *     <ul>A tree cannot contain paths that are a subset of other paths (e.g. foo and foo.bar in V2 should collide and fail)</ul>
+ * </li>
```
```suggestion
 * <ul>
 *     <li>Tree values contain either a nested tree or a field path</li>
 *     <li>A tree cannot contain paths that are a subset of other paths (e.g. foo and foo.bar in V2 should collide and fail)</li>
 * </ul>
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+ * </li>
+ *
+ * See KIP-821.
+ *
```
```suggestion
 * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures">KIP-821</a>
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersion.java:
```diff
@@ -0,0 +1,112 @@
+ * Defines semantics of field paths by versioning.
+ * <p>
+ * See KIP-821.
+ *
```
```suggestion
 * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures">KIP-821</a>
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+ * Invariants:
+ * <li>
+ *     <ul>Tree values contain either a nested tree or a field path</ul>
+ *     <ul>A tree cannot contain paths that are a subset of other paths (e.g. foo and foo.bar in V2 should collide and fail)</ul>
```
If "tree" refers to the `pathTree` field, we shouldn't include detail on it in the Javadoc for this class. This space should be used for information that summarizes the API of the class for callers, so that if, e.g., I see `MultiFieldPaths` in use in an SMT somewhere and don't know what that is, I can get a decent summary by reading up on what's written here instead of having to pore over the source code.

I'm not saying that this information isn't still useful (it definitely helped me understand the values of the `pathTree` map better), just that it should go somewhere else. In this case, probably a comment above the `pathTree` field or in the Javadoc for the `buildPathTree` method.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+public class MultiFieldPaths implements FieldPath {
+
+    final Map<String, Object> pathTree;
+    final List<SingleFieldPath> paths;
```
This field is only used in the `size` method, which is itself unused. We can probably get rid of both.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    Map<String, Object> buildPathTree(List<SingleFieldPath> paths, int stepIdx, Map<String, Object> pathTree) {
+        if (paths.size() == 1) { // optimize for paths with a single member
+            SingleFieldPath path = paths.get(0);
+            if (path != null) {
```
We've already filtered out null elements in the constructor, do we really need this check?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    }
+
+    Map<String, Object> buildPathTree(List<SingleFieldPath> paths, int stepIdx, Map<String, Object> pathTree) {
+        if (paths.size() == 1) { // optimize for paths with a single member
```
Does this optimize anything for processing of data over the lifetime of an SMT, or just for building the path tree when the SMT is being instantiated? If the former then we may want to keep it, but if the latter, we should probably remove it and let the general-case logic handle things in its place.

jeqo replied:
Agree, similar to previous premature optimization.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+            // group paths by prefix
+            final Map<String, List<SingleFieldPath>> groups = new HashMap<>();
+            for (SingleFieldPath path : paths) {
+                if (path != null) {
```
Same question as above; do we really need this check?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+                                        || path.stepAt(stepIdx + 1) == null)) {
+                                    throw new IllegalArgumentException(
+                                            "Path " + other + " and " + path + " are overlapping. "
+                                                    + "Paths need to point to leaf values");
```
Is this correct? What if I'm using the `HeaderFrom` SMT and want to create headers from both a `user` and a `user.email` field? Shouldn't that be permitted?

jeqo replied:
oh haven't think about this one! I agree it's a valid use-case.
This will require keeping "user" on the tree. 

I have gave a try and use empty string to represent "user" on the tree:
- "user"
  - ""
  - "email"

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+                        });
+                    }
+                }
+            }
```
I think this can be simplified using the Java 8 streams API:

```suggestion
            final Map<String, List<SingleFieldPath>> groups = paths.stream()
                    .filter(p -> p.stepAt(stepIdx) != null)
                    .collect(Collectors.groupingBy(
                            path -> path.stepAt(stepIdx)
                    ));
```

If we really need to check to make sure we don't add the same path twice to a value in that map, we can collect into a `Set` instead:
```java
            final Map<String, Set<SingleFieldPath>> groups = paths.stream()
                    .filter(p -> p.stepAt(stepIdx) != null)
                    .collect(Collectors.groupingBy(
                            path -> path.stepAt(stepIdx),
                            Collectors.toSet()
                    ));
```

If we really need to do the check for overlapping paths, we can do that after creating the `groups` field by iterating over each of its values.

jeqo replied:
I like this! I was trying to be cautious of users adding overlapping paths. But by using Sets we could conflate them and only use the children paths.
Giving a try now.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+ * </li>
+ *
+ * See KIP-821.
+ *
```
```suggestion
 * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures">KIP-821</a>
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+ * to access values within a data object (either {@code Struct} or {@code Map<String, Object>}).a
+ * <p>
+ * If the SMT requires accessing multiple fields on the same data object,
+ * use {@code FieldPaths} instead.
```
```suggestion
 * use {@link MultiFieldPaths} instead.
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+ * If the SMT requires accessing multiple fields on the same data object,
+ * use {@code FieldPaths} instead.
+ * <p>
+ * The field path semantics are defined by the syntax version {@code FieldSyntaxVersion}.
```
```suggestion
 * The field path semantics are defined by the {@link FieldSyntaxVersion syntax version}.
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+ * Invariants:
+ * <li>
+ *     <ul>A field path can contain one or more steps</ul>
+ * </li>
```
```suggestion
 * <ul>
 *     <li>A field path can contain one or more steps</li>
 * </ul>
```

C0urante replied:
(Same thought RE moving this to a comment above the `path` field if it's referring to internal implementation logic of the class instead of its externally-facing API)

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+
+    static SingleFieldPath ofV2(String field) {
+        return of(field, FieldSyntaxVersion.V2);
+    }
```
These can be replaced by static utility methods in the necessary test classes

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+    private static final char DOT_CHAR = '.';
+    private static final char BACKSLASH_CHAR = '\\';
+
+    private static final Cache<String, SingleFieldPath> PATHS_CACHE = new SynchronizedCache<>(new LRUCache<>(16));
```
I think this optimization is premature. We only parse paths when configuring SMTs initially, so once we've started processing data (which is the real critical path performance-wise), this doesn't really gain us anything.

jeqo replied:
Agree. At the beginning was thinking paths were going to be built at runtime, but as they are built at construction no much value. Removing it.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+        }
+    }
+
+    private String[] buildFieldPathV2(String pathText) {
```
This can (and probably should) be static

C0urante replied:
Think this is still unaddressed?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+     * @param field potentially containing backticks
+     * @throws IllegalArgumentException when there are incomplete backtick pairs
+     */
+    private String escapeBackticks(String field) {
```
This can (and probably should) be static

C0urante replied:
Also, is `escapeBackticks` really accurate here? That name had me thinking the purpose of the method was to add backslashes to backticks (which is how they're escaped in the field syntax spec), but it seems like the actual purpose is to remove backslashes instead.

jeqo replied:
~would `removeEscapedBackticks` work better?~
Not sure about it neither. The purpose is to remove the backslash used to escape backticks. maybe `processEscapedBackticks` would be a better name?

jeqo replied:
Going for `processEscapedBackticks` let's check it on the next round.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+                    throw new IllegalArgumentException("Unknown syntax version: " + version);
+            }
+        }
+    }
```
I think we can simplify this (assuming `buildFieldPathV2` and `escapeBackticks` are made static, and the cache is eliminated) by making the constructor do nothing except take in and store the `path` field, and putting the parsing and version branching logic all in the `of` method. I also don't think we should silently accept the case where `field` is null, but if this is necessary to accommodate an existing use case, let me know and we can probably keep that part.


```suggestion
    public static SingleFieldPath of(String field, FieldSyntaxVersion version) {
        Objects.requireNonNull(field, "Field path cannot be null");
        switch (version) {
            case V1:
                return new SingleFieldPath(new String[] { field } );
            case V2:
                String[] path = buildFieldPathV2(field);
                return new SingleFieldPath(path);
            default:
                throw new IllegalArgumentException("Unknown syntax version");
        }
    }

    SingleFieldPath(String[] path) {
        this.path = path;
    }
```

jeqo replied:
I like it! Adapting the constructor to this approach

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+    private String[] buildFieldPathV2(String pathText) {
+        // if no dots or wrapping backticks are used, then return path with single step
+        if (!pathText.contains(String.valueOf(DOT_CHAR))) {
+            return new String[] {pathText};
```
If this isn't necessary and can be handled by the general-case logic, we should remove it

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+
+    private static final char BACKTICK_CHAR = '`';
+    private static final char DOT_CHAR = '.';
+    private static final char BACKSLASH_CHAR = '\\';
```
Nit: these would all be a bit more readable without the `_CHAR` suffix (we know by the type that they're all `char`):
```suggestion
    private static final char BACKTICK = '`';
    private static final char DOT = '.';
    private static final char BACKSLASH = '\\';
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+                        // check that it is not escaped or wrapped in another backticks pair
+                        if (idx < s.length() - 1 // not wrapping the whole field path
+                                && (s.charAt(idx + 1) != DOT_CHAR // not wrapping
+                                || s.charAt(idx - 1) == BACKSLASH_CHAR)) { // ... or escaped
```
We can make this a little bit easier to read by pulling out all of the sub-conditions of the `if` statement into their own variables:
```suggestion
                    boolean endOfPath = idx >= s.length() - 1;
                    boolean notWrappingField = !endOfPath && s.charAt(idx + 1) != DOT;
                    boolean escaped = s.charAt(idx - 1) == BACKSLASH;
                    // check that it is not escaped or wrapped in another backticks pair
                    if (!endOfPath && (notWrappingField || escaped)) {
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+                        s.delete(0, s.length());
+                    }
+                }
+            }
```
We can actually do this without creating a `StringBuilder` or mutating strings (beyond using `String::substring`) at all:
```suggestion
        int idx = 0;
        while (idx < pathText.length()) {
            // start processing backtick pair, if any
            if (pathText.charAt(idx) == BACKTICK) {
                idx++;
                final int start = idx; // this is where the "real" (i.e., not-wrapped-by-backticks) field name starts

                // find backtick closing pair
                while (true) {
                    idx = pathText.indexOf(String.valueOf(BACKTICK), idx);
                    if (idx == -1) { // if not found, fail
                        throw new IllegalArgumentException("Incomplete backtick pair in field path: " + pathText);
                    }
                    boolean endOfPath = idx >= pathText.length() - 1;
                    boolean notWrappingField = !endOfPath && pathText.charAt(idx + 1) != DOT;
                    boolean escaped = pathText.charAt(idx - 1) == BACKSLASH;
                    // check that it is not escaped or wrapped in another backticks pair
                    if (!endOfPath && (notWrappingField || escaped)) {
                        idx++; // move index forward and keep searching
                    } else { // it's the closing pair
                        String field = pathText.substring(start, idx);
                        steps.add(escapeBackticks(field));
                        idx += 2; // increment by two (once for the backslash, and one for a potential dot following it)
                        break;
                    }
                }
            } else { // process dots in path
                final int start = idx; // this is where the field name starts
                idx = pathText.indexOf(String.valueOf(DOT), idx);
                if (idx == -1) {
                    // we've reached the end of the path
                    String field = pathText.substring(start);
                    steps.add(escapeBackticks(field));
                    break;
                } else {
                    String field = pathText.substring(start, idx);
                    steps.add(escapeBackticks(field));
                    idx++;
                }
            }
        }
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+                } else { // process dots in path
+                    final int atDot = s.indexOf(String.valueOf(DOT_CHAR));
+                    if (atDot > 0) { // get path step and move forward
+                        steps.add(escapeBackticks(s.substring(0, atDot)));
```
Aren't we guaranteed that the field name has no backticks here? Why the call to `escapeBackticks`?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+                        steps.add(escapeBackticks(s.substring(0, atDot)));
+                        s.delete(0, atDot + 1);
+                    } else { // add all
+                        steps.add(escapeBackticks(s.toString()));
```
Same question RE the call to `escapeBackticks` (FWIW all the unit tests seem to pass when I remove them both)

C0urante reviewed (commented):
Thanks @jeqo! The internal API for dealing with nested fields looks pretty smooth (evidenced by the small diffs in, e.g., the `ExtractField`) class. I've started to review the implementation of that API, but haven't been able to complete a full pass yet, nor have I been able to review most of the SMTs that leverage it. Still, I wanted to give you something to respond to between now and when I'm able to look at the rest. Let me know if you have any questions!

C0urante reviewed (commented):
Thanks @jeqo! The internal API for dealing with nested fields looks pretty smooth (evidenced by the small diffs in, e.g., the `ExtractField`) class. I've started to review the implementation of that API, but haven't been able to complete a full pass yet, nor have I been able to review most of the SMTs that leverage it. Still, I wanted to give you something to respond to between now and when I'm able to look at the rest. Let me know if you have any questions!

C0urante commented:
I'm also wondering about the spec for V2 field syntax. The KIP states that "if the backticks are in a wrapping position (opening or closing a field name), then need to be escaped with backslash", but I think we might actually want something like "if a backtick is followed by a dot in the field name, the backtick should be escaped with a backslash".

For example:

- With a field name of <code>a.b</code>, the user would have to enclose the field in backticks to handle the dot in the name, so they would specify <code>`a.b`</code> in the connector config
- With a field name of <code>`a.b</code>, there's no dot after the backtick in the name, so nothing special is necessary; the user can just add that backtick to the field name in the connector config and use <code>``a.b`</code>
- Similarly, with a field name of <code>a.b`</code>, the user could just specify <code>`a.b``</code> in the connector config
- But, with a field name of <code>a`.b</code>, the user would need to signal to us that the <code>`.</code> part of that name is not the end of the field name, so they would have to specify <code>`a\`.b`</code> in the connector config

Does that work, or is the current logic/spec correct?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,585 @@
+            }
+        }
+        return s.toString();
+    }
```
I took a stab at simplifying this whole method to 1) remove unnecessary `StringBuilder` instantiation, 2) handle all backtick escape logic on the fly (instead of in a follow-up step with the `escapeBackticks` method), and 3) follow the slightly-tweaked spec I outlined in my latest comment. LMKWYT

```suggestion
    private String[] buildFieldPathV2(String pathText) {
        final List<String> steps = new ArrayList<>();
        int idx = 0;
        while (idx < pathText.length() && idx >= 0) {
            if (pathText.charAt(idx) != BACKTICK_CHAR) {
                final int start = idx;
                idx = pathText.indexOf(String.valueOf(DOT_CHAR), idx);
                if (idx > 0) { // get path step and move forward
                    String field = pathText.substring(start, idx);
                    steps.add(field);
                    idx++;
                } else { // add all
                    String field = pathText.substring(start);
                    steps.add(field);
                }
            } else {
                StringBuilder field = new StringBuilder();
                idx++;
                int start = idx;
                while (true) {
                    idx = pathText.indexOf(String.valueOf(BACKTICK_CHAR), idx);
                    if (idx == -1) { // if not found, fail
                        throw new IllegalArgumentException("Incomplete backtick pair in path: " + pathText);
                    }
                    boolean endOfPath = idx >= pathText.length() - 1;
                    if (endOfPath) {
                        field.append(pathText, start, idx);
                        // we've reached the end of the path, and the last character is the backtick
                        steps.add(field.toString());
                        idx++;
                        break;
                    }
                    if (pathText.charAt(idx + 1) != DOT_CHAR) {
                        // this backtick isn't followed by a dot; include it in the field name, but continue
                        // looking for a matching backtick that is followed by a dot
                        idx++;
                        continue;
                    }
                    if (pathText.charAt(idx - 1) == BACKSLASH_CHAR) {
                        // this backtick was escaped; include it in the field name, but continue
                        // looking for an unescaped matching backtick
                        field.append(pathText, start, idx - 1)
                                .append(BACKTICK_CHAR);
                        idx++;
                        start = idx;
                        continue;
                    }
                    // we've found our matching backtick
                    field.append(pathText, start, idx);
                    steps.add(field.toString());
                    idx += 2; // increment by two to include the backtick and the dot after it
                    break;
                }
            }
        }
        return steps.toArray(new String[0]);
    }
```

jeqo commented:
@C0urante thanks for such a generous review! I learnt a lot while applying suggestions.

I think I covered most of them, and let a few open ones with comments to cover them. 

About the last suggestion and refactor, would you mind checking its format? I think GitHub Markdown has messed up the comment and it's hard for me to understand the proposal.

jeqo committed:
KAFKA-14226: chore(connect:transform): Prepare transforms to access field apis

jeqo committed:
KAFKA-14226: feat(connect:transform): Adapt ExtractField SMT to use nested fields

jeqo committed:
KAFKA-14226: feat(connect:transform): Adapt HeaderFrom SMT to use nested fields

jeqo committed:
KAFKA-14226: feat(connect:transform): Adapt MaskField SMT to use nested fields

jeqo committed:
KAFKA-14226: feat(connect:transform): Adapt TimestampConverter SMT to use nested fields

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+
+/**
+ * A FieldPath is composed by one or many field names, known as steps,
+ * to access values within a data object (either {@code Struct} or {@code Map<String, Object>}).a
```
```suggestion
 * A SingleFieldPath is composed of one or more field names, known as steps,
 * to access values within a data object (either {@code Struct} or {@code Map<String, Object>}).
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+ * <p>The field path semantics are defined by the {@link FieldSyntaxVersion syntax version}.
+ *
+ * <p>Paths are calculated once and cached for further access.
+ *
```
We removed caching, right?

```suggestion
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+                        throw new IllegalArgumentException("Incomplete backtick pair in path: " + pathText);
+                    }
+
+                    boolean atEndOfPath = idx >= pathText.length() - 1;
```
The idea with [this suggestion](https://github.com/apache/kafka/pull/12637#discussion_r1148773643) was to simplify the `if` statement because there were multiple boolean conditions that we were evaluating.

That strategy isn't as useful here because each condition is the only thing we evaluate in each corresponding `if` statement, and each condition is only used once.

IMO it'd be more readable to just keep things simple in this method, e.g. `if (idx >= pathText.length() - 1)`.

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Add tests to address [comment]

[comment]: https://github.com/apache/kafka/pull/12637#issuecomment-1485251519

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,499 @@
+    // - Tree values contain either a nested tree or a field path
+    // - A tree can contain paths that are a subset of other paths
+    //   (e.g. foo and foo.bar in V2 would be kept)
+    final Map<String, Object> pathTree;
```
It looks like we've implemented a [trie](https://en.wikipedia.org/wiki/Trie) here. Usually these types of data structures are implemented with recursive types. In this case, we might have a `boolean leaf` field (which is used to track whether the current field should be included in the transformation) and a `Map<String, MultiFieldPaths> children` field that contains all children that share the same prefix, and can be empty if the current node is a leaf node.

This would be a bit cleaner if we could manage it. Do you think you could give it a shot?

jeqo replied:
Sure! Thanks for introducing this structure. I gave it a shot, please check the latest commits and let me know if it's close to what you had in mind.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+            if (pathText.charAt(idx) != BACKTICK) {
+                final int start = idx;
+                idx = pathText.indexOf(String.valueOf(DOT), idx);
+                if (idx > 0) { // get path step and move forward
```
Sorry, I screwed this part up--should be `>=`, not `>`. It's possible (though unlikely) that someone could give us something like `...` and right now that would cause an infinite loop while trying to parse the path.

Also, to catch another similar edge case, we should add this check at the very end of the method, right before the return statement:

```java
if (!pathText.isEmpty() && pathText.charAt(pathText.length() - 1) == DOT)
    steps.add("");
```

I verified this with a new test case in the `SingleFieldPathTest` suite, in case it helps:
```java
@Test
public void testEmptyFieldNames() {
    assertArrayEquals(
            new String[] {"", "", ""},
            new SingleFieldPath("..", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"foo", "", ""},
            new SingleFieldPath("foo..", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"", "bar", ""},
            new SingleFieldPath(".bar.", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"", "", "baz"},
            new SingleFieldPath("..baz", FieldSyntaxVersion.V2).path()
    );
}
```

jeqo replied:
Sure, applying this and adding the test case.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+                    idx += 2; // increment by two to include the backtick and the dot after it
+                    break;
+                }
+
+
```
Nit: unnecessary newlines (allows less code to fit on the screen, makes it a little harder to read)
```suggestion
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+                if (i == path.length - 1) { // get value
+                    return current.field(path[i]);
+                } else { // iterate
+                    current = current.field(path[i]).schema();
```
Won't this cause an NPE if the field doesn't exist in this schema?

jeqo replied:
Yeap, good catch. Adding null validations.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,500 @@
+                if (i == path.length - 1) { // get value
+                    return current.get(path[i]);
+                } else { // iterate
+                    current = current.getStruct(path[i]);
```
Won't this cause a `DataException` if the field isn't found in this struct?

C0urante replied:
Took a stab at this; it's a little unclean but all the tests pass; LMKWYT

```java
public Object valueFrom(Struct struct) {
    Struct current = struct;
    for (String pathSegment : Arrays.asList(path).subList(0, path.length - 1)) {
        // Check to see if the field actually exists
        if (current.schema().field(pathSegment) == null) {
            return null;
        }
        current = current.getStruct(pathSegment);
    }
    return current.get(path[path.length - 1]);
}
```

(It also feels like it might be a little more ergonomic if `path` were a `List` instead of an array, but that's not a blocker)

jeqo replied:
Similar to the trie suggestion, I think a list may be simpler than an array. I gave it a try.

C0urante reviewed (commented):
Thanks @jeqo, this is looking good! Haven't finished a complete round yet but wanted to leave a few more comments.

jeqo committed:
fix(docs): use correct field path name

Co-authored-by: Chris Egerton <fearthecellos@gmail.com>

jeqo committed:
fix(docs): rm comment from old implementation

Co-authored-by: Chris Egerton <fearthecellos@gmail.com>

jeqo committed:
fix(docs): rm unnecessary new line

Co-authored-by: Chris Egerton <fearthecellos@gmail.com>

jeqo committed:
fixup: address review [1]

[1]: https://github.com/apache/kafka/pull/12637#discussion_r1300537859

jeqo committed:
fixup: address review [2]

[2]: https://github.com/apache/kafka/pull/12637#discussion_r1301881456

jeqo committed:
fixup: address review [3]

[3]: https://github.com/apache/kafka/pull/12637#discussion_r1301885093

jeqo committed:
fixup: initial attempt to address [4], without applying suggestion yet

[4]: https://github.com/apache/kafka/pull/12637#discussion_r1301890421

jeqo committed:
fixup: use a backing list instead of array, prep'ng for suggestion [4]

[4]: https://github.com/apache/kafka/pull/12637#discussion_r1301890421

jeqo committed:
fixup: apply suggestion from [4]

[4]: https://github.com/apache/kafka/pull/12637#discussion_r1301890421

jeqo committed:
fixup! make buildFieldPathV2 static

jeqo committed:
fixup! apply suggestion to only handle backslash before dot [1]

[1]: https://github.com/apache/kafka/pull/12637#issuecomment-1485251519

jeqo committed:
fixup! handle null values

jeqo committed:
KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

jeqo committed:
KAFKA-14226: feat(connect:transform): Introduce MultiFieldPaths to apply changes to multiple paths

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce MultiFieldPaths to apply changes to multiple paths

Refactor to use trie abstraction, following [1]

[1]: https://github.com/apache/kafka/pull/12637#discussion_r1301863257

jeqo commented:
Thanks, @C0urante!

About [1], I agree with your proposal. It will make the proposed notation more intuitive and reduce the cognitive load of users. I have updated the KIP and adjusted the representation with your proposal.

[1]: https://github.com/apache/kafka/pull/12637#issuecomment-1485251519

mimaison commented:
@jeqo Are you planning to complete this work? This would be a nice KIP to have.

jeqo commented:
@mimaison thanks for the reminder! I checked with @C0urante and decided to open a new PR to continue this review as the current one just got too long for anyone to grasp. 
Opened a new one: https://github.com/apache/kafka/pull/15379 where we can continue the review -- will close this one.
